src/App.tsx

```diff
@@ -1,26 +1,209 @@
+import { TodoList } from './components/TodoList';
+import { Filter } from './utils/enamFilter';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [filter, setFilter] = useState<'all' | 'active' | 'completed'>(
+    Filter.All,
+  );
+  const [error, setError] = useState('');
+  const [title, setTitle] = useState('');
+  const [loading, setLoading] = useState<number | null>(null);
+  const showError = (message: string) => {
+    setError(message);
+
+    setTimeout(() => {
+      setError('');
+    }, 3000);
+  };
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => showError('Unable to load todos'))
+      .finally(() => setLoading(0));
+  }, []);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const addTodo = (todoTitle: string) => {
+    setLoading(1);
+    setTempTodo({
+      id: 0,
+      userId: 2215,
+      title: todoTitle,
+      completed: false,
+    });
+
+    // Add a new todo
+    postTodo(title.trim())
+      .then((newTodo: Todo) => {
+        setTodos(prevTodos => [...prevTodos, newTodo]);
+        setError('');
+        setTitle('');
+      })
+      .catch(() => showError('Unable to add a todo'))
+      .finally(() => {
+        setLoading(0);
+        setTempTodo(null);
+      });
+  };
+
+  const updateTodoCheck = (id: number) => {
+    setLoading(id);
+
+    const todoToUpdate = todos.find(todo => todo.id === id);
+
+    if (!todoToUpdate) {
+      setLoading(0);
+
+      return;
+    }
+
+    const updatedTodo = {
+      ...todoToUpdate,
+      completed: !todoToUpdate.completed,
+    };
+
+    // Update a todo
+    renewTodo(id, updatedTodo)
+      .then(() => {
+        setTodos(currentTodos =>
+          currentTodos.map(todo =>
+            todo.id === id ? { ...todo, completed: !todo.completed } : todo,
+          ),
+        );
+        setLoading(0);
+      })
+      .catch(() => showError('Unable to update a todo'))
+      .finally(() => setLoading(0));
+  };
+
+  const updateTodoTitle = (
+    event: React.FormEvent,
+    id: number,
+    value: string,
+  ) => {
+    event.preventDefault();
+    setLoading(id);
+
+    const todoToUpdate = todos.find(todo => todo.id === id);
+
+    if (!todoToUpdate) {
+      setLoading(0);
+
+      return;
+    }
+
+    const updatedTodo = {
+      ...todoToUpdate,
+      title: value,
+    };
+
+    // Update a todo
+    renewTodo(id, updatedTodo)
+      .then(() => {
+        setTodos(currentTodos =>
+          currentTodos.map(todo =>
+            todo.id === id ? { ...todo, title: value } : todo,
+          ),
+        );
+        setLoading(0);
+      })
+      .catch(() => showError('Unable to update a todo'))
+      .finally(() => {
+        setLoading(0);
+      });
+    // setEditTitle('');
+  };
+
+  const deleteTodo = (id: number) => {
+    setLoading(id);
+
+    // Remove a todo
```

Comment:
it is not good practice comment function. good code should be "self commented". It means that just reading code should be understandabel what it doing.